Files: phase 3

smotornyuk opened the pull request:
Phase 3 of #8920

This PR consolidates DB models and API actions into 4,000 lines of code.

## Improvements of features introduced in phase 2

* config option to switch between the original unchanged Uploader, which does not use files at all, and the new Uploader that relies on configurable storage. This option is for extensions that extend and heavily override the original uploader and can break after even the slightest change in its logic.

## API 

I want to try a different layout of logic separation inside `ckan.logic`. As new logic belongs to the "file" domain, instead of separating actions between `get`, `delete`, and `update`, I created new files:

* `logic/action/file.py`: file actions
* `logic/auth/file.py`: file auth functions
* `logic/schema/file.py`: file schemas

I added just the most generic actions, like create, delete, etc. Other operations, like resumable uploads and the ability to override a file, may sound tempting, but I see a lot of problems and cannot make a good generic solution for them. For now, I recommend adding custom actions over the existing generic functionality.

And there is a "hidden" `file_search` action. It can be called normally, I just excluded it from documentation, that's why it's "hidden". I'm not sure if searching for files should be available via API(for now, I locked it behind sysadmin permission). And as @amercader is working on new search specs, the syntax of search parameters is also unstable. Even though I'm not going to index the file via a search engine, using similar search parameters is natural. I diverged slightly from the current specification, and here are  the rules:

Soring is controlled by `sort`. It can be either string `sort: "name"`, which means sort by a single column ASC. Alternatively, it can be a list of strings, `sort: ["name", "size"]` to sort by multiple columns ASC simultaneously. Finally, it can be listed with nested 2-element lists, `sort: [["name", "asc"], ["size", "desc"]]`, to control direction. The second and third styles can be mixed.  There is no string processing and `"name desc"` won't be split into column and direction - you must provide list with two elements if you need more control.

Filters are similar to specification from `ckanext-search`, but I decided to use `$` prefix for every operator, not only for `and` / `or`. In addition, `$and` / `$or`, must be specified on the top level, which is similar to the way one builds queries in SQLAlchemy. For example, if you want to find a file with the name `hello` or `world` (and you do not want to use `$in`), you must use the first payload, not the second

```py
# correct
filters = {
  "$or": [
    {"name": "hello"},
    {"name": "world"},
  ]
}

# INCORRECT
filters = {
  "name": {
    "$or": ["hello", "world"] # `$or` is not allowed as field-level operator
  }
}

```

Here's an example of a complex query that is still valid. Note, it's top-level `$or` contains 2 items that are essentially describing the same conditions. The second item just shorter because it relies on implications(e.g. `field: [1,2,3]` implies `filed: {$in: [1,2,3]}`:

```json
{
    "$or": [
        {
            "$and": [
                {"name": {"$in": ["wsgi.py", "README.rst"]}},
                {"size": {"$gt": 3167, "$lt": 3200}},
                {"content_type": {"$eq": "text/plain"}}
            ]
        },
        {
            "name": ["wsgi.py", "README.rst"],
            "size": {"$gt": 3167, "$lt": 3200},
            "content_type": "text/plain"
        }
    ]
}
```
## DB model

There are 3 new models: File, Owner, and OwnerTransferHistory. All of them are defined using a new declarative style of model definition from SQLAlchemy v2.0. It's different from the imperative mapping and classic declarative style, but it integrates well into the typing system, and it's worth switching to it.

`File` contains file details. I'm not sure if `atime` and `mtime`(access and modification time) columns are useful, and can drop them. Other columns are reasonable.

`Owner` model is an extension for file management logic that helps with permissions. The idea is that a file can be either unowned or owned by something that has *type* (user) and *id*(user ID). When a file is accessed, we check the relationship between the owner of the file and the user who accesses the file. And depending on this relationship(the user itself is an owner of the file; the file is owned by the dataset that is created by the user), CKAN decides what the user can do with the file.

`OwnerTransferHistory` is a hidden model. Whenever a file is moved from one owner to another, a new record appears in this table. In this way, we can find out who uploaded the file and who owned it before the current owner. This functionality can be covered by activities, but files are different from public entities that are mentioned inside activities, and I think that keeping ownership history separately is better. Anyway, this model is not promoted and we can drop it in the future if it does not work in the real world.

## CLI

A couple of new CLI commands under `ckan file ...` route. Move file from one storage to another, stream the content of the file to STDOUT, report missing files, show statistics about used space, etc.

## CKANConfig.subtree

New method in the config object that returns options that match given prefix and, optionally, transforms these options into nested dictionary. When you have following config and want all options for sqlalchemy:

```ini
redis.a = 1
redis.b = 2

sqlalchemy.url = 1
sqlalchemy.nested.hello = 2
sqlalchemy.nested.world = 3
```

you can extract either of the following dictionaries:

```py
# flat
{"url": 1, "hested.hello": 2, "nested.world": 3}

# nested
{
  "url": 1, 
  "hested": {
    "hello" : 2, 
    "world": 3,
  },
}
```

I'm using this method to collect storage configuration, but it can be used elsewhere, especially for features that share a common prefix for multiple config options.

## New config options

There are 3 of them, the most important is `ckan.files.owner.cascade_access`, which makes the `Owner` model useful. Documentation of this option explains how ownership works.

## raw storage adapters and CKAN-storage adapters

File-keeper does not require CKAN, which means some features, like converting file content into a Flask response, must be provided by CKAN. Because of it, original storage adapters defined inside file-keeper can be used in CKAN, but it's preferable to wrap them into a class that enables a few CKAN-specific operations. You can find such wrappers under `ckan/lib/files/default/`. 

## Additional adapters

I registered S3, Azure Blob, and GCS adapters in addition to Apache Libcloud. The latter supports a lot of different providers, but it gives access to a very limited set of operations, which is why I think it is worth exposing native adapters for the most popular storage. BTW, file-keeper documentation explains how to [emulate these storages locally](https://datashades.github.io/file-keeper/adapters/emulate/).

There is one unsolved problem, though. Adapters rely on additional dependencies, such as `boto3` or `azure-storage-blob`. These dependencies publish new versions very often, and these versions are backward compatible. Because of this, I don't want to add a fixed version of these libraries to CKAN requirements. And even more, these libraries are not required if you are not going to use the corresponding storage, so I don't want to install them at all by default. 
And here's the problem. I don't want to add packages to requirements, because they will not be used. But if somebody wants to use them, they must find out how the package must be installed. I'm sure that even if I write with big red letters the required dependencies, people won't read it and wonder why the storage is not working.

Any ideas? Maybe, when the user configures storage that is not available due to the missing driver, I should add the recommended `pip install ...` command to the exception text?


## Adapters for tests

I added a storage adapter that keeps files in memory for tests, and another adapter that just ignores all operations without raising an error(like `/dev/null`). 
I can also register a Redis adapter, because the memory adapter does not share files between processes, thus you cannot use it to test CLI or background jobs. But this is a really rare scenario, so I suggest using the filesystem adapter + `/tmp` for it. 


## ContextCache

Cache is stored inside the context. Similar to `context["package"]` that holds **current** package. But it stores items with their primary key, so it's safe to share this cache between actions as long as you invalidate items that were removed from DB(which I do). I need it to reduce the number of DB requests.

## Permissions

This part looks intricate on first glance, but it's actually quite robust and simple. All API actions have auth functions. Every auth function checks if the user has permission, depending on the nature of the operation:

* update (can rename or transfer file to a different owner)
* read (can read file metadata)
* delete (can delete the file)

All permissions depend on the owner. If the file is owned by the package, the user has permission(read/delete/update) for the file if they can do the same thing(read/delete/update) with the package itself. In this way, if a user can only read the dataset, they can download the file, but cannot modify or delete it. If the user cannot see the dataset, they won't see the file as well.

And, if it doesn't work, apart from chaining the auth function, you can use an interface method that extends permission checks. Using it, you can do anything crazy enough, like give a user permission to download the file if the filename matches the user's name.

## Blueprint

A new blueprint with a single route that can be used to download the file. Generic implementation just streams the content, so CKAN works as a proxy between storage and the user. However, this point can be extended on the storage level. Via ~20 lines of code you'll be able to redirect the user to a signed temporal S3 link. And user will download file directly from the cloud, reducing load on CKAN app. 

Maybe I'll add a PR that makes this possible out of the box in the future. Currently, the flow for signed/resumable/multipart uploads is not settled in file-keeper, and I'm trying to avoid unstable features.

smotornyuk commented:
@Zharktas , @EricSoroos just confirmed behavior of S3 adapter. If you install file-keeper with S3 extra (I'm using alpha while this PR is in review):

```sh
pip install 'file-keeper[s3]==0.1.0a4'
``` 

you can connect to S3 using either explicit credentials, which is not recommended, as you mentioned on the call:

```py
from file_keeper import make_storage
storage = make_storage(
  "test", {
    "type": "file_keeper:s3",
    "bucket": "file-keeper",
    "key": "***",
    "secret": "***", 
    # region, etc.
  }
)
```

The second option is to specify `AWS_*` envvars, and then you'll need only the bucket name:
```sh
$ AWS_ACCESS_KEY_ID=*** AWS_SECRET_ACCESS_KEY=*** python

>>> from file_keeper import make_storage
>>> storage = make_storage(
...   "test", {
...     "type": "file_keeper:s3",
...     "bucket": "file-keeper",
...   }
... );
```

And you can also use `~/.aws/credentials`. Assuming you have their default profile like this:
```ini
[default]
aws_access_key_id = ***
aws_secret_access_key = ***
region=***
```
 You can also use just the bucket name to initialize the connection. 

These examples are using the raw `file_keeper:s3` adapter, but the same thing will work for CKAN version of it `ckan:s3` inside an initialized CKAN app

DOes it cover your concerns?

wardi commented on ckan/cli/file.py:
```diff
@@ -0,0 +1,567 @@
+    owner_col = (
+        sa.func.concat(model.Owner.owner_type, " ", model.Owner.owner_id)
+        if verbose
+        else sa.func.concat(model.Owner.owner_type, "")
```
It looks like `.label()` works on orm attributes
```suggestion
        else model.Owner.owner_type
```

wardi commented on ckan/cli/file.py:
```diff
@@ -0,0 +1,567 @@
+        size = fk.humanize_filesize(file.size)
+        click.echo(f"\t{file.id}: {file.name} [{file.content_type}, {size}]")
+
+    if remove and click.confirm("Do you want to delete these files?"):
```
good to have a `yes` option here too for consistency

smotornyuk replied:
I followed your advice above and replaced all confirmations with command flags. So you either specify the flag like `--remove` and files will be removed, or run command without this flag and see only statistics(with a message that explains, how they can be removed)

smotornyuk commented on ckan/cli/file.py:
```diff
@@ -0,0 +1,567 @@
+        dest = open(output, "wb")  # noqa: SIM115
+
+    for chunk in content_stream:
+        click.echo(chunk, nl=False, file=dest)
```
Yes, `.write` is definitely better here, I overlooked this moment

smotornyuk commented on ckan/cli/file.py:
```diff
@@ -0,0 +1,567 @@
+                if file := model.Session.scalar(model.File.by_location(item, src)):
+                    info.into_object(file)
+                    file.storage = dest
+                    model.Session.commit()
```
How about storages that contains thousands of files and just few problematic records. If you copied 500 files and 501th caused `ExistingFileError`, the operation is interrupted and you won't be able to continue the process, even after removing the 501th, because you already have yet another 500 matching files. So the only option would be to manually specify the rest 500 files if you want to continue after solving the problem. Or remove previously copied files and repeat operation from the start.

Initially I thought that for the user it would be to do as much work as possible and then collect reported filenames, to solve the problem for them manually. But after your comment, I think that the best option is to let user decide the strategy of error handling. So I'll add 2 new parameters `skip-missing-files` and `skip-existing-files`. By default, the whole operation is interrupted when error happens, but if user enables these flags, corresponding errors will be reported and operation continues.

wardi replied:
I meant to say exit with an error code outside of the for loop, not interrupting the progress part-way.

wardi replied:
i.e. keep a flag that tracks if there were any failures then at the exit of the function make sure the process exits with an error code

smotornyuk commented on ckan/common.py:
```diff
@@ -133,6 +133,60 @@ def get(self, key: str, default: Any = SENTINEL) -> Any:
 
         return super().get(key, default)
 
+    def subtree(self, prefix: str = "", depth: int = 0, keep_prefix: bool = False):
```
Yes, probably it's better to transform this into an utility function and keep config object as close to `dict` as possible, without extra methods. But I moved it outside of declarations module, because it is not bound logically to declarations - it's just filters and transforms dictionary content and can be ponentially used with any dictionary, not only config. That's why I don't want to keep it next to declarations, but cannot find any better place for it.

smotornyuk commented on ckan/config/config_declaration.yaml:
```diff
@@ -1403,6 +1403,111 @@ groups:
+            Do not use configurable storages for Uploader and ResourceUploader,
+            keeping previous implementation of these classes. Enable this
+            option to keep backward compatibility with extensions that extend
+            storage logic.
```
`bool` options are always boolean, so default value is `False`, even when it's not specified. As for warning, I added details inside 8745th changelog record - not sure if it should be duplicated here, because people only need to keep it in mind during migration, when they are taking decision to opt-out from new features.

wardi commented on ckan/lib/files/__init__.py:
```diff
@@ -0,0 +1,196 @@
+#         if table is not None and table.name == owner_model:
+#             return model.Session.get(cls, owner_id)
+
+#     log.warning("Unknown owner type %s with ID %s", owner_type, owner_id)
```
pls remove commented-out code (above, below)

smotornyuk replied:
Done

wardi commented on ckan/lib/files/base.py:
```diff
@@ -0,0 +1,352 @@
+    >>>     ) -> FileData:
+    >>>         reader = upload.hashing_reader()
+    >>>         with open(location, "wb") as dest:
+    >>>             dest.write(reader.read())
```
it is only example docs, but if you follow this the complete file contents will be loaded into memory

smotornyuk replied:
[Replaced with a for-loop](https://github.com/ckan/ckan/pull/9026/changes/5a6cda4a2ad75f1695e05db17fff1dae100b9d0d#diff-cb3b6230c4e2783718964aa0aa3d8ce4d148abeaf22c3b6e7dd80592f2206db3R71-R75) that reads fragments of the file. I don't put much detail about hashing_reader here, because it's up to the user to either rely on the documented features of file-keeper, or implement their own hashing logic.

wardi commented on ckan/lib/files/base.py:
```diff
@@ -0,0 +1,352 @@
+    >>>         self, data: FileData, extras: dict[str, Any]
+    >>>     ) -> bool:
+    >>>         os.remove(data.location)
+    >>>         return True
```
What's the use of the return value if failures result in exceptions?

smotornyuk replied:
Certain storages may return false when the file does not exist, and true when it does. There is no requirement to cause an exception if the file is missing - it's a decision of the driver's author. In file-keeper's documentation, I recommend [returning False](https://datashades.github.io/file-keeper/api/#file_keeper.Storage.remove) when the file does not exist instead of an exception

wardi commented on ckan/lib/files/base.py:
```diff
@@ -0,0 +1,352 @@
+
+    supported_types: list[str] = dataclasses.field(default_factory=list)
+    """Supported types of uploads"""
+    max_size: int = 0
```
Does max_size 0 mean unlimited? (If so I'm not a fan, I'd prefer None or if we really need an int for some reason something like -1)

smotornyuk replied:
I prefer to avoid multiple types for the same value. [Let's choose `-1`](https://github.com/ckan/ckan/pull/9026/changes/5a6cda4a2ad75f1695e05db17fff1dae100b9d0d#diff-cb3b6230c4e2783718964aa0aa3d8ce4d148abeaf22c3b6e7dd80592f2206db3R234) to stay with `int`. And it matches the behavior of built-in functions that have concept of numeric restriction

wardi commented on ckan/lib/files/base.py:
```diff
@@ -0,0 +1,352 @@
+            "Space-separated list of full MIME types, or just type/subtype part."
+        ).set_example("text/csv pdf application video jpeg")
+
+        declaration.declare_bool(key.override_existing).set_description(
```
"overwrite" existing? override is usually only for things like config values

smotornyuk replied:
Any suggestions? I don't mind other options, but cannot choose among `rewrite`, `replace`, `force` and other similar words

wardi replied:
I think "overwrite" would be better than "override" here

smotornyuk replied:
Updated

wardi commented on ckan/lib/files/base.py:
```diff
@@ -0,0 +1,352 @@
+        )
+
+        declaration.declare_bool(key.initialize).set_description(
+            "Prepare storage backend for uploads(create path, bucket, DB). This op",
```
This op ...?

smotornyuk replied:
Accidentally removed the continuation of the line. Restored it

wardi commented on ckan/lib/files/base.py:
```diff
@@ -0,0 +1,352 @@
+            "List of transformations applied to the file location."
+            "\nDepending on the storage type, sanitizing the path or removing"
+            " special characters can be sensible.\nEmpty value leaves location"
+            " unchanged, `uuid` transforms location into random UUID,"
```
there are different flavors of uuid, if it's random does this mean uuid4? If so just say `uuid4`. uuid7 is also a really useful option because they guarantee no conflicts but can also be sorted by creation date/time.

smotornyuk replied:
I'm using v4 - updated as per your suggestion. v7 will not be provided out of the box because it requires python v3.14. But there is a way to register custom transformers(interface method), so individual portals will be able to use it

wardi commented on ckan/lib/files/base.py:
```diff
@@ -0,0 +1,352 @@
+        """Make Flask response with file attachment.
+
+        By default, files are served as attachments and are downloaded as
+        result.
```
the code below says they're sent inline based on the content type

smotornyuk replied:
Extended the docstring explaining how inline/attachment is affected by configuration

wardi commented on ckan/lib/files/base.py:
```diff
@@ -0,0 +1,352 @@
+        resp.headers.set(
+            "content-disposition",
+            disposition,
+            filename=filename or data.location,
```
If we have the content type, shouldn't we include it in the headers returned?

smotornyuk replied:
I like this idea. Also including content-length and checking for redirect, to return it immediately instead of setting extra headers.

wardi commented on ckan/lib/files/default/__init__.py:
```diff
@@ -0,0 +1,34 @@
+    from .gcs import GoogleCloudStorage
+
+with contextlib.suppress(ImportError):
+    from .s3 import S3Storage
```
This sort of import pattern can slow down startup based only on what libraries are in the venv. We could instead import these only in the places they are used so that startup time is based on the config not what python packages are installed

wardi replied:
I would like to see core plugins for each of the storage types, e.g.
`plugins = files_opendal files_s3` means that only the s3 and opendal adapters are available and they import all their dependencies on startup (including erroring out and preventing ckan from starting if any of them are missing)

This also has the side effect of advertising which storage adapters are in use on ckan sites by checking the list of plugins in `status_show`, which I think is really useful information for statistics and debugging.

smotornyuk replied:
This is no longer an issue for core, because all optional plugins are extracted into ckanext-file-keeper-cloud extension. And there I'll apply your suggestions

wardi commented on ckan/lib/uploader.py:
```diff
@@ -404,3 +412,264 @@ def upload(self, id: str, max_size: int = 10) -> None:
                 pass
+
+
+class FkUpload(object):
```
if We're shortening "file keeper upload" let's use "FKUpload" so that it doesn't look like we're shortening a different word.

smotornyuk replied:
:) Noted

wardi commented on ckan/lib/uploader.py:
```diff
@@ -404,3 +412,264 @@ def upload(self, id: str, max_size: int = 10) -> None:
+                 old_filename: Optional[str] = None) -> None:
+        ''' Setup upload by creating a subdirectory of the storage directory
+        of name object_type. old_filename is the name of the file in the url
+        field last time'''
```
what does "last time" mean here?

smotornyuk replied:
[Some code that was added 13 year ago](https://github.com/ckan/ckan/commit/bb6cb7d0ff41). I rephrased this docstring

wardi commented on ckan/lib/uploader.py:
```diff
@@ -404,3 +412,264 @@ def upload(self, id: str, max_size: int = 10) -> None:
+                object_type,
+            )
+
+    def update_data_dict(self, data_dict: dict[str, Any], url_field: str,
+                         file_field: str, clear_field: str) -> None:
+        ''' Manipulate data from the data_dict.  url_field is the name of the
+        field where the upload is going to be. file_field is name of the key
+        where the FieldStorage is kept (i.e the field where the file data
+        actually is). clear_field is the name of a boolean field which
+        requests the upload to be deleted.  This needs to be called before
+        it reaches any validators'''
```
I know this is being compatible with existing code, but we could really use a cleaner interface here.

- the function name is bad because it doesn't say what the function actually does
- the way this function modifies the passed dict in-place is questionable
- the fact that this function needs to run before validators is surprising because users could pass anything easily leading to server errors if this code isn't really careful

wardi replied:
Do we need to keep this interface for maintaining compatibility with existing uploader extensions? Is running this method before validators to allow support for things like direct uploads to a cloud provider?

smotornyuk replied:
This method is used in a couple of public extensions (like ckanext-s3filestore), and my main concern is about my own modifications of portals - previously, this `update_data_dict` was the main method to modify data when you don't need to replace the uploader, but only want to modify it a bit. And I believe that changing this exact method will be the most harmful for compatibility.

And I think that in the future we won't need the Uploader class at all, because files are managed separately, and we should be able to attach a file to anything using a validation schema. Even if you want to do something with file inside the action, you can use a storage object directly, so Uploader will be removed in v3.0, most likely. Refactoring this method now will create more friction for people, because this logic will be used only during one release.

wardi commented on ckan/lib/uploader.py:
```diff
@@ -404,3 +412,264 @@ def upload(self, id: str, max_size: int = 10) -> None:
+            if self.upload_field_storage.filename:
+                self.filename = self.upload_field_storage.filename
+                self.filename = str(datetime.datetime.utcnow()) + self.filename
+                self.filename = munge.munge_filename_legacy(self.filename)
```
`munge_filename_legacy` is the function that replaces all non-ascii characters, which is really bad for many users.

smotornyuk replied:
Yes, bad decision, but replacing it can break existing image paths according to its documentation:

>     Unfortunately it mangles any path or filename extension, so is deprecated.
>     It needs to remain unchanged for use by group_dictize() and
>     Upload.update_data_dict() because if this routine changes then group images
>     uploaded previous to the change may not be viewable.

wardi commented on ckan/lib/uploader.py:
```diff
@@ -404,3 +412,264 @@ def upload(self, id: str, max_size: int = 10) -> None:
+                data_dict[url_field] = self.filename
+
+        # keep the file if there has been no change
+        elif self.old_filename and not self.old_filename.startswith('http'):
```
`self.old_filename.startswith('http')` is not how to check if a filename is a url

smotornyuk replied:
Replaced it with more reliable check as this change won't break existing extensions

wardi commented on ckan/lib/uploader.py:
```diff
@@ -404,3 +412,264 @@ def upload(self, id: str, max_size: int = 10) -> None:
+            if self.clear and self.url == self.old_filename:
+                data_dict[url_field] = ''
+
+    def upload(self, max_size: int = 2) -> None:
```
why `max_size = 2`?

smotornyuk replied:
[Historical reasons](https://github.com/ckan/ckan/commit/bb6cb7d0ff41). Generally, this value will be specified directly when the method is called, so this default value does not affect existing code. I don't want to change or modify it, because extensions may expect this default value

wardi commented on ckan/lib/uploader.py:
```diff
@@ -404,3 +412,264 @@ def upload(self, id: str, max_size: int = 10) -> None:
+                    self.upload_file,
+                )
+            except (files.exc.LargeUploadError, files.exc.ExistingFileError) as err:
+                raise logic.ValidationError({"upload": [str(err)]})
```
```suggestion
                raise logic.ValidationError({self.file_field: [str(err)]})
```

wardi commented on ckan/lib/uploader.py:
```diff
@@ -404,3 +412,264 @@ def upload(self, id: str, max_size: int = 10) -> None:
+                declared_mimetype
+                and allowed_mimetypes
+                and allowed_mimetypes[0] != "*"
+                and declared_mimetype not in allowed_mimetypes
```
the mimetype checking here is very different than `is_supported_type` above. We should have a consistent way these comparisons are done to avoid unpleasant surprises.

wardi replied:
I favor keeping the strict checking like this over the `is_supported_type` implementation

smotornyuk replied:
I replaced this complex condition by `is_supported_type`, but changed the logic of this function to be as strict as replaced block. No partial and fuzzy matches except for a single `*`.

wardi commented on ckan/lib/uploader.py:
```diff
@@ -404,3 +412,264 @@ def upload(self, id: str, max_size: int = 10) -> None:
+            self.filename = munge.munge_filename(self.upload_file.filename)
+            resource['url'] = self.filename
+            resource['url_type'] = 'upload'
+            resource['last_modified'] = datetime.datetime.utcnow()
```
As above, I'd really prefer we had a clearer, simpler interface if possible.

It's very surprising to have the `resource` dict parameter values modified after constructing a `FkResourceUpload` instance.

smotornyuk replied:
Agree, but this must be kept intact for compatibility. If we are going to remove this code eventually, it's better not to force users to make extra work that will require removal in v3.0

wardi commented on ckan/logic/__init__.py:
```diff
@@ -1000,3 +1006,50 @@ def index_remove_package(id_: str):
+
+
+class ContextCache:
+    """Cache for storing and retrieving values in the context."""
```
`fresh_context` was written to avoid this pattern. We shouldn't be using the context dict as a way to pass values up and down the call stack because that leads to bugs that are very difficult to understand and reproduce.

What's the specific feature that needs to pass values this way and why?

smotornyuk replied:
Removed this class completely. As you suggested, I'm using SQLAlchemy's identity map and `auth_user_obj` from context to reduce amount of SQLqueries

amercader commented on ckan/logic/action/file.py:
```diff
@@ -0,0 +1,775 @@
+    """
+    logic.check_access("file_search", context, data_dict)
+
+    sess = context["session"]
```
Let's use `session` for readability and consistency

amercader commented on ckan/logic/action/file.py:
```diff
@@ -0,0 +1,775 @@
+
+        owner.owner_id = data_dict["owner_id"]
+        owner.owner_type = data_dict["owner_type"]
+        owner.pinned = data_dict["pin"]
```
We are setting this a few lines below

smotornyuk replied:
[Updated](https://github.com/ckan/ckan/blob/files-phase-3/ckan/logic/action/file.py#L711-L712)

amercader commented on ckan/logic/action/file.py:
```diff
@@ -0,0 +1,775 @@
+
+@logic.side_effect_free
+@logic.validate(schema.file_search)
+def file_search(  # noqa: C901, PLR0912, PLR0915
```
I have now seen that `file_owner_scan` is the only place that uses `file_search` and just passes two basic filters so it makes sense to not expose the action just yet.

amercader commented on ckan/config/config_declaration.yaml:
```diff
@@ -1403,6 +1403,111 @@ groups:
+          ``package_show``, can see all files owned by the package; anyone who
+          passes ``package_update``/``resource_update`` can modify files owned
+          by the package/resource; anyone who passes
+          ``package_delete``/``resource_delete`` can delete files owned by the
```
A fairly common requirement is to have public metadata (i.e. allow `package_show` or `resource_show`) but restricted downloads. Would be good to have an example on how to do that with the new system (not here, in the docs)

amercader commented on ckan/lib/files/default/null.py:
```diff
@@ -0,0 +1,26 @@
+from __future__ import annotations
```
This storage is not documented. What are it's use cases?

smotornyuk replied:
just as memory storage, this one is intended for tests, that do not need to access the content of uploaded files. Forgot to document it

wardi commented on ckan/logic/action/file.py:
```diff
@@ -0,0 +1,775 @@
+    else is controlled by storage. The same file may be accepted by one storage
+    and rejected by another, depending on its configuration.
+
+    The action is way too powerful to use it directly. The recommended approach
```
```suggestion
    This action is not intended to be used directly. The recommended approach
```

wardi replied:
Maybe with explanation like "because files created this way will be considered orphans/unlinked files"

smotornyuk replied:
Yes, I added even more details to the docstring of this action. It has pretty impressive volume already

wardi commented on ckan/logic/action/file.py:
```diff
@@ -0,0 +1,775 @@
+log = logging.getLogger(__name__)
+
+
+def _set_user_owner(context: Context, item_type: str, item_id: str):
```
Instead of passing a required parameter in the context dict, pass `user` as a normal parameter here

smotornyuk replied:
After all changes this extra function does not seem so valuable. I inlined it into the only two places where it was used

wardi commented on ckan/logic/action/file.py:
```diff
@@ -0,0 +1,775 @@
+    "$in": "IN",
+    "$is": "IS",
+    "$isnot": "IS NOT",
+    "$like": "LIKE",
+    "$ilike": "ILIKE",
+}
+
+
+def _process_field(col: Column[Any], value: Any):  # noqa: C901
+    """Transform `{"field":{"$eq":"value"}}` into SQL filters."""
+    if isinstance(value, list):
+        value = {"$in": value}
+
+    elif value is None:
+        value = {"$eq": None}
+
+    elif not isinstance(value, dict):
+        value = {"$eq": value}
+
+    for operator, filter in value.items():
+        column = col
+        if operator not in _op_map:
+            raise logic.ValidationError(
+                {"filters": [f"Operator {operator} is not supported"]}
+            )
+        op = _op_map[operator]
+        if filter is None:
+            if op == "=":
+                op = "is"
+            elif op == "!=":
+                op = "IS NOT"
+
+        elif operator == "$in" and isinstance(filter, list):
+            filter = tuple(filter)
+        elif not isinstance(filter, col.type.python_type):
+            filter = str(filter)
+            column = sa.func.cast(column, sa.Text)
+
+        func = column.bool_op(op)
+        yield func(filter)
```
when https://github.com/ckan/ckan/pull/9028 is merged we can share same filters.py implementation from @amercader

smotornyuk replied:
Yes, I definitely lookin for this.

wardi commented on ckan/logic/action/file.py:
```diff
@@ -0,0 +1,775 @@
+
+    .. note:: Requires storage with `CREATE` capability.
+
+    :param name: human-readable name of the file.
```
IIUC this value must be unique within the storage, and that is typically in the format `path/filename.ext`, we should make this clear in the docstring

smotornyuk replied:
Extended the documentation. Filenames are sanitized, so nested paths are not allowed by default. I added an explanation with example, how this can be achieved

wardi commented on ckan/logic/action/file.py:
```diff
@@ -0,0 +1,775 @@
+            raise logic.ValidationError({"upload": [msg]})
+        data_dict["name"] = filename
+
+    filename = secure_filename(data_dict["name"])
```
looking at the werkzeug docs this might not be an ideal function to use. It's limited to ascii-only, removes paths and may result in an empty filename.

smotornyuk replied:
Didn't know about these details. I replacing `secure_filename` with CKAN's `munge_filename`

wardi commented on ckan/logic/action/file.py:
```diff
@@ -0,0 +1,775 @@
+    try:
+        storage_data = storage.analyze(data_dict["location"])
+    except files.exc.MissingFileError:
+        raise logic.NotFound("file")
```
```suggestion
        raise logic.NotFound("file not found at location")
```

wardi commented on ckan/logic/action/file.py:
```diff
@@ -0,0 +1,775 @@
+        raise logic.NotFound("file")
+
+    fileobj = model.File(
+        name=secure_filename(storage_data.location),
```
transforming the location with `secure_filename` means that it might not be found by the next call to File.by_location, right?

smotornyuk replied:
No. We are transforming the location to use it as a _name of this DB record_. Name is intended for the public-facing interface. For example, we can use it when showing all files on the webpage or when the user downloads the file. It's different from the actual _location_ of the file in the storage controlled by the `location` field. So we can transform `name` however we like(here I'm simplifying it because location can contain path segments if the upload process was customized) and it will not change the actual location of the file, nor our ability to find it.

wardi commented on ckan/logic/action/file.py:
```diff
@@ -0,0 +1,775 @@
+    logic.check_access("file_search", context, data_dict)
+
+    sess = context["session"]
+    columns = dict(**model.File.__table__.c, **model.Owner.__table__.c)
```
I think it's worth listing the columns explicitly here, otherwise when something gets added to the Model it will magically change the search interface which could be surprising.

smotornyuk replied:
Done

wardi commented on ckan/logic/action/file.py:
```diff
@@ -0,0 +1,775 @@
+    try:
+        total: int = sess.scalar(stmt.with_only_columns(sa.func.count()))  # pyright: ignore[reportAssignmentType]
+    except ProgrammingError:
+        sess.rollback()
```
need to rollback a select?

smotornyuk replied:
Yes, probably has to do something with our auto-transaction settings. You can try it yourself:
```
# ckan shell
import sqlalchemy as sa

# works
model.Session.scalar(sa.select(sa.text('1')))

# now cause a problem
model.Session.scalar(sa.select(sa.column('none')))

# does not work anymore, untill you rollback the transaction
model.Session.scalar(sa.select(sa.text('1')))
```

wardi commented on ckan/logic/action/file.py:
```diff
@@ -0,0 +1,775 @@
+    fileobj = cache.get_model("file", data_dict["id"], model.File)
+
+    if not fileobj:
+        raise logic.NotFound("file")
```
```suggestion
        raise logic.NotFound("file id not found")
```

wardi commented on ckan/logic/action/file.py:
```diff
@@ -0,0 +1,775 @@
+) -> ActionResult.FileOwnerScan:
+    """List files of the owner.
+
+    .. warning:: This action has hight probability to be changed or removed in
```
```suggestion
    .. warning:: This action has a high probability to be changed or removed in
```

wardi replied:
Is this isn't used, is there a reason to keep this action?

smotornyuk replied:
I think we really should build an interface that shows all files of the user, database, resource, group, etc. But there are a lot of questions that can be solved only after using the file in the real world and seeing usage patterns. That's the part I left for implementation after v2.12 release.

But while we will work on the file management UI only after v2.12, I think that some portals would also like to implement this logic by themselves. That's why I added `owner_scan`, to at least start from the same direction.

wardi commented on ckan/logic/auth/file.py:
```diff
@@ -0,0 +1,300 @@
+    owner_id: str,
+    operation: types.FileOwnerOperation,
+) -> bool:
+    """Decide if user is allowed to perform operation on the owner."""
```
based on the owner?

smotornyuk replied:
Extended docs to make it more clear

wardi commented on ckan/model/file.py:
```diff
@@ -0,0 +1,164 @@
+        storage (str): storage that contains the file
+        content_type (str): MIMEtype
+        size (int): size in bytes
+        hash (str): checksum
```
different back ends will support one or more different hashing algorithms. We could store the algorithm with a prefix on the string, I didn't immediately see a standard for this but there likely is one.

wardi replied:
e.g. DCAT 3 supports multiple checksums with different algorithms: https://www.w3.org/TR/vocab-dcat-3/#Class:Checksum

smotornyuk replied:
I added a separate field to File model and FileData dataclass that called `algorithm` and it stores the name of checksum algorithm. Somehow I don't want to puth both the algorithm and the hash into the same string. As a bonus, there is a new config option `hashing_algorithm` that can be used to modify the algorithm of the backend. But it can be ignored by certain storages(like all cloud storages at the moment), which is mentioned in config declaration.

I'll try to enable configurable algorithm for cloud providers in future versions of file keeper.

wardi commented on ckan/model/owner.py:
```diff
@@ -0,0 +1,132 @@
+
+
+@registry.mapped_as_dataclass
+class OwnerTransferHistory:
```
This kind of thing might be better left to the activity plugin. Does any core file functionality depend on the OwnerTransferHistory?

smotornyuk replied:
Nothing depends on it, but I think that tracking the file's origin is quite important. While activities **can** be used for it, their permission labels are not stabilized yet, and it seems that they work only with datasets.

When activities fully support access control, this model can be migrated into activities, but till that moment, I'd rather track these details in this way than leave them untracked.

wardi commented on ckan/model/owner.py:
```diff
@@ -0,0 +1,132 @@
+
+
+@registry.mapped_as_dataclass
+class Owner:
```
+1 on changing this to FileOwner and dropping the owner_type, or maybe even better roll this into the File model and save all the extra work required managing separate entities.

wardi commented on ckan/logic/converters.py:
```diff
@@ -254,3 +257,55 @@ def remove_whitespace(value: Any, context: Context) -> Any:
+
+
+def parse_filesize(value: str | int):
+    """Convert human-readable filesize into an integer."""
```
If we're exporting this as a validator this should include examples of what values can be parsed and tests covering those examples just in case the underlying file_keeper implementation changes at some point.

Converting from a human-readable file size to an integer loses important precision information, what do we need this converter for?

smotornyuk replied:
Added few examples and tests.

This validator is used by `max_size` option of storage to avoid uncertainty when specifying filesizes. FOr example, CKAN core uses [megabytes](https://github.com/ckan/ckan/blob/master/ckan/config/config_declaration.yaml#L1394-L1404) for resource and image sizes, while xloader uses [bytes](https://github.com/ckan/ckanext-xloader/blob/master/ckanext/xloader/config_declaration.yaml#L45). The latter is not readable, but supports precise size control The former is not clear without documentation and does not allow to set exact fractions of megabyte, but is readable.

New validator gives user exactly what is specified in configuration: `1B`, `1MB`, `10MiB` (yes, it knows difference between megabytes and mebibites), `1GB` etc. There are not questions about units of measure with these values when you see them.

As for precision, I'm not sure, because it supports any integral value of bytes. I.e, if you want half of MB/MiB, you can set `500KB`/`512KiB` (btw, letter case does not matter). So there is no loss as long as you are not using floating-point.

wardi replied:
I was concerned that it would be used for a user-supplied metadata field at some point and if someone writes `5MiB` then it would be stored as `5242880` which looks like a very specific value and another user might not realize that it was a "round" number.

But, given that it's intended for a max_size option and not validating a metadata field this is fine.

wardi commented on ckan/model/file.py:
```diff
@@ -0,0 +1,164 @@
+        hash (str): checksum
+        ctime (datetime): date of creation
+        mtime (datetime | None): date of the last update
+        atime (datetime | None): date of last access(unstable)
```
Is `atime` desirable as a default feature?

`atime` turns read operations (someone accessing a file) into write operations on ckan's db (updating the atime) which could have a significant performance impact for busy sites. This might be better as a feature added by an extension, ideally one that uses redis or a separate table in the db to store/update access times so that access logging has a lighter impact on the application.

smotornyuk replied:
Yes, both mtime and atime are stubs that I'm considering for now. Ideally, I want files to be immutable, because it solves millions of problems and I cannot see a solid reason to **replace** existing file instead of creating a new one and use it's identifier. So mtime is unlikely to ever be used. 
access time is something that can be easily extracted in some storage types(like local filesystem), that's why I initially added it. But to use it universally, we must call `touch` method to update this property(yes, you are right, at the moment I'm not calling it at all) whenever file is accessed. While visiting `/download` endpoint initially sounded like a good solution, I suspect that there may be more ways to access the file in future, and in the end we'll just mess up this property, forgetting to update it somewhere.
So removing both mtime and atime seems the best solution for now. If anyone really needs these properties, he can create a wrapper class around the storage class/API/`/download` that modifies access time when it has most sense.

wardi commented on ckan/model/file.py:
```diff
@@ -0,0 +1,164 @@
+        compare=False,
+    )
+
+    def dictize(self, context: Any) -> dict[str, Any]:
```
use normal parameters instead of passing in `context`

smotornyuk replied:
Done

wardi commented on ckan/tests/cli/test_file.py:
```diff
@@ -0,0 +1,267 @@
+
+
+@pytest.fixture
+def with_temporal_storage(
```
```suggestion
def with_temp_storage(
```
temporal means related to time :smile:

smotornyuk replied:
I lived in delusion all those years

wardi commented on ckan/tests/logic/auth/test_file.py:
```diff
@@ -0,0 +1,396 @@
+
+# operation == "file_transfer" and config["ckan.files.owner.transfer_as_update"]
+# operation == "file_scan" and config["ckan.files.owner.scan_as_update"]
+
```
```suggestion
```

wardi commented on ckan/tests/logic/auth/test_file.py:
```diff
@@ -0,0 +1,396 @@
+
+
+class TestPermissionDownload(BasePermission):
+    action = "permission_download_file"
```
is it better to use parameterize than inheritance?

smotornyuk replied:
Yes, I was aggressively copying and pasting code and overlooked this place

wardi commented on ckan/views/file.py:
```diff
@@ -0,0 +1,73 @@
+#     if isinstance(storage, files.Storage):
+#         return storage.as_response(data)
+
+#     return base.abort(422, "File is not downloadable")
```
please remove commented view

smotornyuk replied:
Have time to experiment more with views, and now I have removed commented code and added 3 generic download views that can satisfy the majority of situations when a download is required:
* `download` - used to download files created via API. Relies on the file ID and the ownership system to understand whether the user is allowed to download the file.
* `trusted_download` uses a temporal JWT token to download any file. In this way, you can build features like sharing private files or sending private files to a moderator for review. Generally, this one can be used if the ownership system is too strict
* `public_download` is available for storages with `public = true`. In this way, we can give access to group images to everyone in a storage-agnostic way.

All views are using the `as_response` method of the storage that ideally redirects the user to the cloud or serves the file in an efficient manner. In the worst case, if there is no efficient way to serve the file provided by the storage adapter(if it's stored inside DB), CKAN works as a proxy and streams the file, just like it does for datastore dump.

wardi commented on doc/api/index.rst:
```diff
@@ -444,6 +444,14 @@ ckan.logic.action.delete
+
+.. automodule:: ckan.logic.action.file
+   :members:
+   :exclude-members: file_search
```
don't document file_search?

smotornyuk replied:
Initially I excluded it from documentation because it wasn't stabilized. But now, as Adria suggested, I removed this action completely and using private function instead

wardi commented on ckan/cli/file.py:
```diff
@@ -0,0 +1,567 @@
+                    model.Session.commit()
+
+
+@storage.command("clean")
```
maybe `remove-all-files` just to be 100% understood?

smotornyuk replied:
Renamed to `remove-files`

smotornyuk commented on ckan/config/config_declaration.yaml:
```diff
@@ -1403,6 +1403,111 @@ groups:
+          .. note: Do not add ``user`` without specific storage to this
+            list. Files may be temporarily owned by user during resource
+            creation. Using cascade access rules with ``user`` exposes such
+            temporal files to anyone who can read user's profile.
```
Never thought about it, it's really nice idea. I'll see how we can achieve this.

smotornyuk commented on ckan/lib/files/default/memory.py:
```diff
@@ -0,0 +1,22 @@
+from __future__ import annotations
```
I created it for tests, when we need to access content in future, but forgot to document it

smotornyuk commented on ckan/lib/files/default/opendal.py:
```diff
@@ -0,0 +1,42 @@
+from __future__ import annotations
```
I added this one, because it supports a bigger number of storages than libcloud. But as people will mostly use S3/Azure/GCS, I think that we can drop it. I'll keep this adapter available inside [ckanext-files](https://github.com/DataShades/ckanext-files/blob/master/ckanext/files/storage/opendal.py) if anyone ever want to use it.

wardi commented on ckan/config/declaration/__init__.py:
```diff
@@ -203,7 +203,14 @@ def load_core_declaration(self):
+        # dynamic information computed depending on the list of registered
+        # storages. It's still a part of core declarations and must remain in
+        # this method to be discoverable via config CLI.
+        loader(self, "files")
```
Just a thought since the meeting:
- we're moving the cloud providers to an extension, so would we need dynamic declarations at all if we have separate plugins for each supported back end?

smotornyuk replied:
[Extended the comment](https://github.com/ckan/ckan/blob/files-phase-3/ckan/config/declaration/__init__.py#L209-L224) with mention that this logic can be extracteed into generic method if another use of dynamic options appear

wardi commented on ckan/cli/file.py:
```diff
@@ -0,0 +1,567 @@
+
+    file_data = files.FileData.from_object(file)
+    if not start and end is None and storage.supports(files.Capability.STREAM):
+        content_stream = storage.stream(file_data)
```
Is this a limitation of filekeeper?

wardi commented on ckan/model/file_owner.py:
```diff
@@ -0,0 +1,132 @@
+    )
+
+    item_id: Mapped[text] = mapped_column(primary_key=True)
+    item_type: Mapped[text] = mapped_column(primary_key=True)
```
Since a file can only have one owner I still think it's cleaner to merge this with the file model.

But if FileOwner stays separate we can at least remove `item_type` and make `item_id` a FK to the File table so that postgresql protects us from bugs in our code leading to invalid data.

smotornyuk replied:
I want to keep the owner separately, as it's rather an additional feature, which is not required by the file system per se. Files can work without owners, and if we see that this ownership system does not work for us, it will be easier to deal with it when it is kept separately. 
I like this strict ownership because it eliminates a lot of edge cases, but I want to leave some freedom in decision-making, whether this system works for particular portals. That's why I put a lot of work into the permission system and made it predictable, but completely configurable and replaceable at the same time.

As for keeping only file_id and making it a foreign key - it's a good idea. Applied this change

smotornyuk commented:
Tests are added to extensions - they will turn green when CKAN is published.

I tried to publish the extension but I don't see any option to add CKAN organization as an owner/collaborator to the project. Here's my collaborators tab from pypi. In `user` field I can select only user accounts:
<img width="1177" height="883" alt="image" src="https://github.com/user-attachments/assets/11ca8b50-b996-4e83-abe4-2211ef0aa2a9" />


UPD: I sent you an owner invitation

wardi commented:
Thank you @smotornyuk .
There are an additional 47 comments I added (as part of a review? I don't know how github reviews work) that don't have replies or associated changes. Would you take another look?

<img width="1091" height="1660" alt="Screenshot from 2026-04-01 10-33-17" src="https://github.com/user-attachments/assets/f6b67296-eadc-4117-bf4a-4190a3b9f57e" />

smotornyuk commented:
I suspect you left this comment using "start a review". They are invisible for everyone unless you submit this review:
<img width="840" height="698" alt="image" src="https://github.com/user-attachments/assets/15be6e90-3a48-4762-b119-aca16e7deebf" />

smotornyuk commented:
A generic question for your review: In the screenshot, I see 3 questions: `munge legacy`, `startswith http`, `max_size = 2`. For all these cases, I didn't write this code, but just kept existing lines to be as close as possible to the original behavior.  Do you think we should change these places? 

I suspect that after v2.12, when we complete the migration to the new file management and start using file IDs instead of names, most of these checks will be removed. For example, in the case of group image, we can use(at the moment it's a mere suggestion, I haven't yet found the ideal way to manage files) a new field `image_id` that will contain the ID of the file with the group image. When it's non-empty, we can use this ID to generate the value of the `image_url` field. When it's empty, we assume there are no uploads and allow `image_url` to be provided manually. That's why `startswith('http')` will naturally disappear in the future. Do we need to change such non-ideal places for v2.12 just to get rid of them eventually? Keeping them in this non-ideal state will simplify backports and fixes if we discover an issue related to these lines - we'll be able to provide same fix for v2.12 and v2.11

smotornyuk commented:
@wardi , I think that I addressed all issues. It's your turn now(I can add few more commits when tests finish and I see if any of them fail)

smotornyuk commented:
Fixed tests. If you see `ckan/tests/controllers/test_group.py::TestGroupSearch::test_group_search_results` failing, you can ignore it. This issue is caused by flaky test and I'll provide a solution in as separate PR

smotornyuk commented:
I updated docs of ckanext-file-keeper-cloud and added more details of using S3 with resoruce and group uploads. It highlight the difference between public and private uploads and focuses on the fact that bucket must be correctly configured. And it suggests few tricks to handle misconfigured bucket if you have no permission to modify it's settings. Check last two sections ["use cloud storage for resource uploads" and "use cloud storage for user and group uploads"](https://github.com/ckan/ckanext-file-keeper-cloud#use-cloud-storage-for-resource-uploads)
And in the beginning of ["Available adapters"](https://github.com/ckan/ckanext-file-keeper-cloud#use-cloud-storage-for-resource-uploads) I emphasised the danger of using inline credentials in the config file
